lib/game_runner.rb

```diff
@@ -1,37 +1,43 @@
 require_relative "grid"
 require_relative "printer"
+require_relative "player_input"
+require_relative "grid_printer"
 
 class GameRunner
-  attr_reader :stdout, :stdin, :grid, :printer
+  attr_reader :stdout, :stdin, :grid, :printer, :player_input, :grid_printer
 
   def initialize(stdout, stdin, grid = Grid.new)
     @stdout = stdout
     @stdin = stdin
     @grid = grid
-  end
-
-  def printer
-    @printer ||= Printer.new(stdout, grid)
+    @printer ||= Printer.new(stdout)
+    @grid_printer = GridPrinter.new(stdout, grid)
+    @player_input = PlayerInput.new(stdin, grid, printer)
   end
 
   def run
-    printer.print_welcome_message
-    demand_valid_coordinates
+    print_welcome_message
+    print_grid
+    get_valid_coordinate
+    mark_grid
+    print_grid
   end
 
   private
 
-  def demand_valid_coordinates
-    while coordinate_invalid?
-      printer.print_coordinates_error
-    end
+  def print_welcome_message
+    printer.print_welcome_message
+  end
+
+  def print_grid
+    grid_printer.print_grid
   end
 
-  def coordinate_invalid?
-    !grid.coordinate_valid?(get_user_input)
+  def get_valid_coordinate
+    player_input.get_valid_coordinate
```

Comment:
Even though we are getting the coordinate somehow, it isn't returned. That is kind of confusing. I would expect something named `get...` to return a value.